#📌 1 - EXTRAÇÃO

1.1 - EXTRAÇÃO DOS DADOS

In [103]:
import pandas as pd
import requests

url = "https://raw.githubusercontent.com/ingridcristh/challenge2-data-science/main/TelecomX_Data.json"

response = requests.get(url)
data = response.json()

df = pd.json_normalize(data)


#🔧 2 - TRANSFORMAÇÃO

2.1 - CONHECENDO O DATASET

2.1.1 - Ver colunas e tipos

In [104]:
# Exibe as primeiras linhas do DataFrame para uma visão geral inicial dos dados
df.head()

customerID Churn customer.gender  customer.SeniorCitizen customer.Partner  \
0  0002-ORFBO    No          Female                       0              Yes   
1  0003-MKNFE    No            Male                       0               No   
2  0004-TLHLJ   Yes            Male                       0               No   
3  0011-IGKFF   Yes            Male                       1              Yes   
4  0013-EXCHZ   Yes          Female                       1              Yes   

  customer.Dependents  customer.tenure phone.PhoneService phone.MultipleLines  \
0                 Yes                9                Yes                  No   
1                  No                9                Yes                 Yes   
2                  No                4                Yes                  No   
3                  No               13                Yes                  No   
4                  No                3                Yes                  No   

  internet.InternetService  ... internet.OnlineBackup  \
0                      DSL  ...                   Yes   
1                      DSL  ...                    No   
2              Fiber optic  ...                    No   
3              Fiber optic  ...                   Yes   
4              Fiber optic  ...                    No   

  internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
0                        No                  Yes                  Yes   
1                        No                   No                   No   
2                       Yes                   No                   No   
3                       Yes                   No                  Yes   
4                        No                  Yes                  Yes   

  internet.StreamingMovies account.Contract account.PaperlessBilling  \
0                       No         One year                      Yes   
1                      Yes   Month-to-month                       No   
2                       No   Month-to-month                      Yes   
3                      Yes   Month-to-month                      Yes   
4                       No   Month-to-month                      Yes   

  account.PaymentMethod account.Charges.Monthly  account.Charges.Total  
0          Mailed check                    65.6                  593.3  
1          Mailed check                    59.9                  542.4  
2      Electronic check                    73.9                 280.85  
3      Electronic check                    98.0                1237.85  
4          Mailed check                    83.9                  267.4  

[5 rows x 21 columns]

In [105]:
# Retorna a quantidade de linhas e colunas do DataFrame
df.shape

(7267, 21)

In [106]:
# Lista os nomes de todas as colunas do DataFrame
df.columns

Index(['customerID', 'Churn', 'customer.gender', 'customer.SeniorCitizen',
       'customer.Partner', 'customer.Dependents', 'customer.tenure',
       'phone.PhoneService', 'phone.MultipleLines', 'internet.InternetService',
       'internet.OnlineSecurity', 'internet.OnlineBackup',
       'internet.DeviceProtection', 'internet.TechSupport',
       'internet.StreamingTV', 'internet.StreamingMovies', 'account.Contract',
       'account.PaperlessBilling', 'account.PaymentMethod',
       'account.Charges.Monthly', 'account.Charges.Total'],
      dtype='object')

In [107]:
# Mostra o tipo de dado de cada coluna (numérico, texto, booleano, etc.)
df.dtypes

customerID                    object
Churn                         object
customer.gender               object
customer.SeniorCitizen         int64
customer.Partner              object
customer.Dependents           object
customer.tenure                int64
phone.PhoneService            object
phone.MultipleLines           object
internet.InternetService      object
internet.OnlineSecurity       object
internet.OnlineBackup         object
internet.DeviceProtection     object
internet.TechSupport          object
internet.StreamingTV          object
internet.StreamingMovies      object
account.Contract              object
account.PaperlessBilling      object
account.PaymentMethod         object
account.Charges.Monthly      float64
account.Charges.Total         object
dtype: object

In [108]:
# Apresenta um resumo completo do DataFrame: tipos, valores não nulos e uso de memória
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer.gender            7267 non-null   object 
 3   customer.SeniorCitizen     7267 non-null   int64  
 4   customer.Partner           7267 non-null   object 
 5   customer.Dependents        7267 non-null   object 
 6   customer.tenure            7267 non-null   int64  
 7   phone.PhoneService         7267 non-null   object 
 8   phone.MultipleLines        7267 non-null   object 
 9   internet.InternetService   7267 non-null   object 
 10  internet.OnlineSecurity    7267 non-null   object 
 11  internet.OnlineBackup      7267 non-null   object 
 12  internet.DeviceProtection  7267 non-null   object 
 13  internet.TechSupport       7267 non-null   objec

2.1.2. - Abrir o dicionario de dados

In [109]:
dic_url = "https://raw.githubusercontent.com/ingridcristh/challenge2-data-science/main/TelecomX_dicionario.md"
print(requests.get(dic_url).text)

#### Dicionário de dados

* `customerID`: número de identificação único de cada cliente
* `Churn`: se o cliente deixou ou não a empresa 
* `gender`: gênero (masculino e feminino) 
* `SeniorCitizen`: informação sobre um cliente ter ou não idade igual ou maior que 65 anos 
* `Partner`:  se o cliente possui ou não um parceiro ou parceira
* `Dependents`: se o cliente possui ou não dependentes
* `tenure`:  meses de contrato do cliente
* `PhoneService`: assinatura de serviço telefônico 
* `MultipleLines`: assisnatura de mais de uma linha de telefone 
* `InternetService`: assinatura de um provedor internet 
* `OnlineSecurity`: assinatura adicional de segurança online 
* `OnlineBackup`: assinatura adicional de backup online 
* `DeviceProtection`: assinatura adicional de proteção no dispositivo 
* `TechSupport`: assinatura adicional de suporte técnico, menos tempo de espera
* `StreamingTV`: assinatura de TV a cabo 
* `StreamingMovies`: assinatura de streaming de filmes 
* `Contract`: tipo de co

2.2 - VERIFICANDO INCONSISTÊNCIA NOS DADOS

2.2.1 – Valores ausentes (missing values).

In [110]:
# Se há campos nulos/vazios por coluna
df.isna().sum()


customerID                   0
Churn                        0
customer.gender              0
customer.SeniorCitizen       0
customer.Partner             0
customer.Dependents          0
customer.tenure              0
phone.PhoneService           0
phone.MultipleLines          0
internet.InternetService     0
internet.OnlineSecurity      0
internet.OnlineBackup        0
internet.DeviceProtection    0
internet.TechSupport         0
internet.StreamingTV         0
internet.StreamingMovies     0
account.Contract             0
account.PaperlessBilling     0
account.PaymentMethod        0
account.Charges.Monthly      0
account.Charges.Total        0
dtype: int64

2.2.2 - Verificar se há linhas duplicadas

In [111]:
# Se existem registros (linhas) repetidos.
df.duplicated().sum()

np.int64(0)

2.2.3 = Tipos de dados incompatíveis

In [112]:
# colunas que deveriam ser numéricas/datas, mas vieram como object
df.dtypes

customerID                    object
Churn                         object
customer.gender               object
customer.SeniorCitizen         int64
customer.Partner              object
customer.Dependents           object
customer.tenure                int64
phone.PhoneService            object
phone.MultipleLines           object
internet.InternetService      object
internet.OnlineSecurity       object
internet.OnlineBackup         object
internet.DeviceProtection     object
internet.TechSupport          object
internet.StreamingTV          object
internet.StreamingMovies      object
account.Contract              object
account.PaperlessBilling      object
account.PaymentMethod         object
account.Charges.Monthly      float64
account.Charges.Total         object
dtype: object

2.2.4 - Inconsistências em categorias

In [113]:
# se as colunas categóricas têm valores fora do esperado
df.nunique()

customerID                   7267
Churn                           3
customer.gender                 2
customer.SeniorCitizen          2
customer.Partner                2
customer.Dependents             2
customer.tenure                73
phone.PhoneService              2
phone.MultipleLines             3
internet.InternetService        3
internet.OnlineSecurity         3
internet.OnlineBackup           3
internet.DeviceProtection       3
internet.TechSupport            3
internet.StreamingTV            3
internet.StreamingMovies        3
account.Contract                3
account.PaperlessBilling        2
account.PaymentMethod           4
account.Charges.Monthly      1585
account.Charges.Total        6531
dtype: int64

2.2.4.1 - Avaliando cada coluna

In [114]:
df['customerID'].unique()

array(['0002-ORFBO', '0003-MKNFE', '0004-TLHLJ', ..., '9992-UJOEL',
       '9993-LHIEB', '9995-HOTOH'], dtype=object)

In [115]:
df['Churn'].unique()

array(['No', 'Yes', ''], dtype=object)

In [116]:
df['customer.gender'].unique()

array(['Female', 'Male'], dtype=object)

In [117]:
df['customer.SeniorCitizen'].unique()

array([0, 1])

In [118]:
df['customer.Partner'].unique()

array(['Yes', 'No'], dtype=object)

In [119]:
df['customer.Dependents'].unique()

array(['Yes', 'No'], dtype=object)

In [120]:
df['customer.tenure'].unique()

array([ 9,  4, 13,  3, 71, 63,  7, 65, 54, 72,  5, 56, 34,  1, 45, 50, 23,
       55, 26, 69, 11, 37, 49, 66, 67, 20, 43, 59, 12, 27,  2, 25, 29, 14,
       35, 64, 39, 40,  6, 30, 70, 57, 58, 16, 32, 33, 10, 21, 61, 15, 44,
       22, 24, 19, 47, 62, 46, 52,  8, 60, 48, 28, 41, 53, 68, 51, 31, 36,
       17, 18, 38, 42,  0])

In [121]:
df['phone.PhoneService'].unique()

array(['Yes', 'No'], dtype=object)

In [122]:
df['phone.MultipleLines'].unique()

array(['No', 'Yes', 'No phone service'], dtype=object)

In [123]:
df['internet.InternetService'].unique()

array(['DSL', 'Fiber optic', 'No'], dtype=object)

In [124]:
df['internet.OnlineSecurity'].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [125]:
df['internet.OnlineBackup'].unique()

array(['Yes', 'No', 'No internet service'], dtype=object)

In [126]:
df['internet.DeviceProtection'].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [127]:
df['internet.TechSupport'].unique()

array(['Yes', 'No', 'No internet service'], dtype=object)

In [128]:
df['internet.StreamingTV'].unique()

array(['Yes', 'No', 'No internet service'], dtype=object)

In [129]:
df['internet.StreamingMovies'].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [130]:
df['account.Contract'].unique()

array(['One year', 'Month-to-month', 'Two year'], dtype=object)

In [131]:
df['account.PaperlessBilling'].unique()

array(['Yes', 'No'], dtype=object)

In [132]:
df['account.PaymentMethod'].unique()

array(['Mailed check', 'Electronic check', 'Credit card (automatic)',
       'Bank transfer (automatic)'], dtype=object)

In [133]:
df['account.Charges.Monthly'].unique()

array([65.6 , 59.9 , 73.9 , ..., 91.75, 68.8 , 67.85])

In [134]:
df['account.Charges.Total'].unique()

array(['593.3', '542.4', '280.85', ..., '742.9', '4627.65', '3707.6'],
      dtype=object)

2.3 - TRATANDO AS INCONSISTÊNCIAS

2.3.1 - Tratamento do Churn (valor vazio)

In [135]:
# Contagem de valores da coluna Churn (incluindo ausentes)
df['Churn'].value_counts(dropna=False)


Churn
No     5174
Yes    1869
        224
Name: count, dtype: int64

In [136]:
# transformar '' em valor ausente (NaN)
import numpy as np

df['Churn'] = df['Churn'].replace('', np.nan)
df['Churn'].isna().sum()

np.int64(224)

2.3.2 - Converter para numérico (account.Charges.Total)

In [137]:
# transformar valores numéricos que estão como texto (object) em número, e também vazio e invalidos.
df['account.Charges.Total'] = pd.to_numeric(
      df['account.Charges.Total'],
          errors='coerce'
          )
df['account.Charges.Total'].isna().sum()


np.int64(11)

2.4 - COLUNA DE CONTAS DIÁRIAS

In [139]:
# Criar coluna "Contas_Diarias" - Valor diário por cliente
df['Contas_Diarias'] = (df['account.Charges.Monthly'] / 30).round(2)
df.head()

customerID Churn customer.gender  customer.SeniorCitizen customer.Partner  \
0  0002-ORFBO    No          Female                       0              Yes   
1  0003-MKNFE    No            Male                       0               No   
2  0004-TLHLJ   Yes            Male                       0               No   
3  0011-IGKFF   Yes            Male                       1              Yes   
4  0013-EXCHZ   Yes          Female                       1              Yes   

  customer.Dependents  customer.tenure phone.PhoneService phone.MultipleLines  \
0                 Yes                9                Yes                  No   
1                  No                9                Yes                 Yes   
2                  No                4                Yes                  No   
3                  No               13                Yes                  No   
4                  No                3                Yes                  No   

  internet.InternetService  ... internet.DeviceProtection  \
0                      DSL  ...                        No   
1                      DSL  ...                        No   
2              Fiber optic  ...                       Yes   
3              Fiber optic  ...                       Yes   
4              Fiber optic  ...                        No   

  internet.TechSupport internet.StreamingTV internet.StreamingMovies  \
0                  Yes                  Yes                       No   
1                   No                   No                      Yes   
2                   No                   No                       No   
3                   No                  Yes                      Yes   
4                  Yes                  Yes                       No   

  account.Contract account.PaperlessBilling account.PaymentMethod  \
0         One year                      Yes          Mailed check   
1   Month-to-month                       No          Mailed check   
2   Month-to-month                      Yes      Electronic check   
3   Month-to-month                      Yes      Electronic check   
4   Month-to-month                      Yes          Mailed check   

  account.Charges.Monthly account.Charges.Total  Contas_Diarias  
0                    65.6                593.30            2.19  
1                    59.9                542.40            2.00  
2                    73.9                280.85            2.46  
3                    98.0               1237.85            3.27  
4                    83.9                267.40            2.80  

[5 rows x 22 columns]

2.5 - PADRONIZAÇÃO E TRANSFORMAÇÃO DE DADOS

2.5.1 - Padronizar texto (Consistência)

In [141]:
# remover espaços e padronizar texto
cat_cols = df.select_dtypes(include="object").columns
df[cat_cols] = df[cat_cols].apply(lambda s: s.str.strip().str.lower())

In [142]:
df.head()

customerID Churn customer.gender  customer.SeniorCitizen customer.Partner  \
0  0002-orfbo    no          female                       0              yes   
1  0003-mknfe    no            male                       0               no   
2  0004-tlhlj   yes            male                       0               no   
3  0011-igkff   yes            male                       1              yes   
4  0013-exchz   yes          female                       1              yes   

  customer.Dependents  customer.tenure phone.PhoneService phone.MultipleLines  \
0                 yes                9                yes                  no   
1                  no                9                yes                 yes   
2                  no                4                yes                  no   
3                  no               13                yes                  no   
4                  no                3                yes                  no   

  internet.InternetService  ... internet.DeviceProtection  \
0                      dsl  ...                        no   
1                      dsl  ...                        no   
2              fiber optic  ...                       yes   
3              fiber optic  ...                       yes   
4              fiber optic  ...                        no   

  internet.TechSupport internet.StreamingTV internet.StreamingMovies  \
0                  yes                  yes                       no   
1                   no                   no                      yes   
2                   no                   no                       no   
3                   no                  yes                      yes   
4                  yes                  yes                       no   

  account.Contract account.PaperlessBilling account.PaymentMethod  \
0         one year                      yes          mailed check   
1   month-to-month                       no          mailed check   
2   month-to-month                      yes      electronic check   
3   month-to-month                      yes      electronic check   
4   month-to-month                      yes          mailed check   

  account.Charges.Monthly account.Charges.Total  Contas_Diarias  
0                    65.6                593.30            2.19  
1                    59.9                542.40            2.00  
2                    73.9                280.85            2.46  
3                    98.0               1237.85            3.27  
4                    83.9                267.40            2.80  

[5 rows x 22 columns]

2.5.2 - Converter Sim/Não para binário (1/0)

In [143]:
# Mapear yes/no → 1/0 usando minúsculas:
map_yn = {'yes': 1, 'no': 0}

cols_yn = [
    'customer.Partner',
    'customer.Dependents',
    'phone.PhoneService',
    'account.PaperlessBilling'
]

for c in cols_yn:
    df[c] = df[c].map(map_yn)


In [144]:
df[cols_yn].head()


customer.Partner  customer.Dependents  phone.PhoneService  \
0                 1                    1                   1   
1                 0                    0                   1   
2                 0                    0                   1   
3                 1                    0                   1   
4                 1                    0                   1   

   account.PaperlessBilling  
0                         1  
1                         0  
2                         1  
3                         1  
4                         1

2.5.3 - Renomear colunas (opcional, clareza)

In [153]:
df = df.rename(columns={
      'customer.gender': 'customer_gender',
      'customer.SeniorCitizen': 'customer_senior',
      'customer.Partner': 'customer_partner',
      'customer.Dependents': 'customer_dependents',
      'customer.tenure': 'customer_tenure',

      'phone.PhoneService': 'phone_service',
      'phone.MultipleLines': 'phone_multiple_lines',

      'internet.InternetService': 'internet_service',
      'internet.OnlineSecurity': 'internet_online_security',
      'internet.OnlineBackup': 'internet_online_backup',
      'internet.DeviceProtection': 'internet_device_protection',
      'internet.TechSupport': 'internet_tech_support',
      'internet.StreamingTV': 'internet_streaming_tv',
      'internet.StreamingMovies': 'internet_streaming_movies',

      'account.Contract': 'account_contract',
      'account.PaperlessBilling': 'account_paperless_billing',
      'account.PaymentMethod': 'account_payment_method',
      'account.Charges.Monthly': 'charges_monthly',
      'account.Charges.Total': 'charges_total',
      'Contas_Diarias': 'contas_diarias'
})

In [154]:
df.columns

Index(['customerID', 'Churn', 'customer_gender', 'customer_senior',
       'customer_partner', 'customer_dependents', 'customer_tenure',
       'phone_service', 'phone_multiple_lines', 'internet_service',
       'internet_online_security', 'internet_online_backup',
       'internet_device_protection', 'internet_tech_support',
       'internet_streaming_tv', 'internet_streaming_movies',
       'account_contract', 'account_paperless_billing',
       'account_payment_method', 'charges_monthly', 'charges_total',
       'contas_diarias'],
      dtype='object')

#📊 3 - CARGA E ANÁLISE

3.1 - ANÁLISE DESCRITIVA

3.2 - DISTRIBUIÇÃO DA EVASÃO

3.3 - CONTAGEM DE EVASÃO POR VARIÁVEIS CATEGÓRICAS

3.4 - CONTAGEM DE EVASÃO POR VARIÁVEIS NUMÉRICAS

3.5 - ANÁLISE DE CORRELAÇÃO ENTRE VARIÁVEIS

#📄4 - RELATÓRIO FINAL